### PR TITLE
raftstore: poll the newer peer list to check stale peer

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: Liqi Geng <gengliqiii@gmail.com>

### What problem does this PR solve?

Problem Summary:
A stale peer will not be destroyed in such conditions
1. Be isolated
2. Be the last removed peer in its peer list
3. Then its region merges to another region.

In this case, PD can not give any help because this region has merged and some related data has been deleted.
After isolation disappears, this peer will send vote msg to other peers in its peer list. But the other peers will ignore this msg and they still think this peer is valid because it exists in their peer list.
This PR introduces a new method to figure out the problem.

### What is changed and how it works?

What's Changed:
Add two new extra message type. `MsgCheckStalePeer` and `MsgCheckStalePeerResponse`.
`MsgCheckStalePeer` is split from `MsgRegionWakeUp`.
The peer can get the newer peer list from other tombstone peers. Then it can use this newer peer list to check if it is stale.

### Related changes

- Need to cherry-pick to the release branch

Tests 

- Integration test

Side effects

No.

### Release note 
Fix an issue that a peer may not be destroyed after isolation recovery.

### CODE REVIEW

**Comment by BusyJay** [comment]
Can you explain more what's exactly the one more method in both title and description?

**Comment by gengliqi** [comment]
> Can you explain more what's exactly the one more method in both title and description?

Done.

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -947,12 +947,21 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
     fn on_extra_message(&mut self, msg: &RaftMessage) {
         let extra_msg = msg.get_extra_msg();
         match extra_msg.get_type() {
-            ExtraMessageType::MsgRegionWakeUp => {
-                self.reset_raft_tick(GroupState::Ordered);
+            ExtraMessageType::MsgRegionWakeUp | ExtraMessageType::MsgCheckStalePeer => {
+                if self.fsm.group_state == GroupState::Idle {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:951]: Why the change?

Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:951]: In one test, I find the leader looks stopped because the `chech-peer-stale-tick` is too frequent and sends too much `MsgCheckStalePeer` to others which causes others `register_raft_base_tick` so I add a check here to avoid useless wakeup.

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -2593,9 +2599,9 @@ impl Peer {
         }
     }
 
-    pub fn bcast_wake_up_message<T: Transport>(&self, trans: &mut T) {
-        let region = self.raft_group.store().region();
-        for peer in region.get_peers() {
+    pub fn bcast_wake_up_message<T: Transport, C>(&self, ctx: &mut PollContext<T, C>) {
+        ctx.need_flush_trans = true;
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2603]: What if peers is empty?

File: components/raftstore/src/store/peer.rs [hunk 3]
```diff
@@ -2618,6 +2624,53 @@ impl Peer {
...
+        }
+    }
+
+    pub fn on_check_stale_peer_response<T: Transport, C>(
+        &mut self,
+        ctx: &mut PollContext<T, C>,
+        check_conf_ver: u64,
+        check_peers: &[metapb::Peer],
+    ) {
+        if self.check_stale_conf_ver < self.region().get_region_epoch().get_conf_ver() {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2664]: This seems unnecessary.

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -947,12 +947,21 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+                }
             }
             ExtraMessageType::MsgWantRollbackMerge => {
                 unimplemented!("remove this after #6584 merged")
             }
+            ExtraMessageType::MsgCheckStalePeerResponse => {
+                self.fsm.peer.on_check_stale_peer_response(
+                    &mut self.ctx,
+                    msg.get_region_epoch().get_conf_ver(),
+                    extra_msg.get_check_peers(),
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:962]: How about `take_check_peers`?

File: components/raftstore/src/store/peer.rs (hunk 3 above)
Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:2664]: Because the check-stale-peer msg should use the latest peer list.

File: components/raftstore/src/store/fsm/peer.rs (hunk 4 above)
Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs:962]: Using reference is enough.

Comment (BusyJay) [reply components/raftstore/src/store/fsm/peer.rs:962]: Then why clone it inside the method?

File: components/raftstore/src/store/peer.rs (hunk 2 above)
Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:2603]: It's another case that can not be handled by this PR.

File: components/raftstore/src/store/fsm/peer.rs [hunk 5]
```diff
@@ -944,8 +944,9 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         Ok(())
     }
 
-    fn on_extra_message(&mut self, msg: &RaftMessage) {
-        let extra_msg = msg.get_extra_msg();
+    fn on_extra_message(&mut self, mut msg: RaftMessage) {
+        let region_epoch = msg.get_region_epoch().clone();
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:948]: Why clone?

File: components/raftstore/src/store/peer.rs [hunk 6]
```diff
@@ -2600,11 +2600,12 @@ impl Peer {
     }
 
     pub fn bcast_wake_up_message<T: Transport, C>(&self, ctx: &mut PollContext<T, C>) {
-        ctx.need_flush_trans = true;
         for peer in self.region().get_peers() {
             if peer.get_id() == self.peer_id() {
                 continue;
             }
+            ctx.need_flush_trans = true;
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2607]: I think it should be set only send success.

Comment (gengliqi) [reply components/raftstore/src/store/peer.rs]: Done.

File: components/raftstore/src/store/peer.rs [hunk 7]
```diff
@@ -2615,9 +2620,59 @@ impl Peer {
...
+        check_peers: Vec<metapb::Peer>,
+    ) {
+        if self.check_stale_conf_ver < self.region().get_region_epoch().get_conf_ver() {
+            self.check_stale_conf_ver = self.region().get_region_epoch().get_conf_ver();
+            self.check_stale_peers = self.region().get_peers().to_vec();
+        }
+        if self.check_stale_conf_ver < check_conf_ver {
+            self.check_stale_conf_ver = check_conf_ver;
+            self.check_stale_peers = check_peers;
+            self.bcast_check_stale_peer_message(ctx);
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:2673]: This may cause a ping pong loop. Suppose `B` is destroyed with newer configuration, and `C` in shutdown. Once `A` sends check message, it will keep ping pong `B` until `C` is started and may cause high CPU usage.

File: components/raftstore/src/store/fsm/peer.rs [hunk 8]
```diff
@@ -1023,9 +1034,9 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         {
             let mut need_gc_msg = util::is_vote_msg(msg.get_message());
             if msg.has_extra_msg() {
-                // A learner can't vote so it sends the wake-up msg to others to find out whether
+                // A learner can't vote so it sends the check-stale-peer msg to others to find out whether
                 // it is removed due to conf change or merge.
-                need_gc_msg |= msg.get_extra_msg().get_type() == ExtraMessageType::MsgRegionWakeUp
+                need_gc_msg |= msg.get_extra_msg().get_type() == ExtraMessageType::MsgCheckStalePeer
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/peer.rs:1039]: Maybe also check `MsgRegionWakeUp`?

Comment (gengliqi) [reply components/raftstore/src/store/fsm/peer.rs]: Yes.
